engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/gateway/InclusiveGatewayTest.java

```diff
@@ -0,0 +1,1340 @@
+        .containsExactlyInAnyOrder("task1", "task2");
+    assertThat(taskEvents.get(0).getKey()).isNotEqualTo(taskEvents.get(1).getKey());
+  }
+
+  @Test
+  public void shouldMergeInclusiveBranchesTwo() {
+    // given
+    final String processId = Strings.newRandomValidBpmnId();
+    final BpmnModelInstance processDefinition =
+        Bpmn.createExecutableProcess(processId)
+            .startEvent("start")
+            .inclusiveGateway("fork")
+            .sequenceFlowId("flow1")
+            .inclusiveGateway("join")
+            .endEvent("end")
+            .moveToNode("fork")
+            .sequenceFlowId("flow2")
+            .manualTask("m2")
+            .connectTo("join")
+            .moveToNode("fork")
+            .sequenceFlowId("flow3")
+            .manualTask("m3")
+            .connectTo("join")
+            .done();
+    // given
+    ENGINE.deployment().withXmlResource(processDefinition).deploy();
+
+    // when
+    final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(processId).create();
+
+    // then
+    final List<Record<ProcessInstanceRecordValue>> events =
+        RecordingExporter.processInstanceRecords()
+            .withProcessInstanceKey(processInstanceKey)
+            .limitToProcessInstanceCompleted()
+            .asList();
+
+    assertThat(events)
+        .extracting(e -> e.getValue().getElementId(), Record::getIntent)
+        .containsSubsequence(
+            tuple("flow1", ProcessInstanceIntent.SEQUENCE_FLOW_TAKEN),
+            tuple("join", ProcessInstanceIntent.ELEMENT_ACTIVATING))
+        .containsSubsequence(
+            tuple("flow2", ProcessInstanceIntent.SEQUENCE_FLOW_TAKEN),
+            tuple("join", ProcessInstanceIntent.ELEMENT_ACTIVATING))
+        .containsSubsequence(
+            tuple("flow3", ProcessInstanceIntent.SEQUENCE_FLOW_TAKEN),
+            tuple("join", ProcessInstanceIntent.ELEMENT_ACTIVATING))
+        .containsOnlyOnce(tuple("join", ProcessInstanceIntent.ELEMENT_ACTIVATING));
+  }
+
+  @Test
+  public void shouldMergeInclusiveBranchesThree() {
+    // given
+    final String processId = Strings.newRandomValidBpmnId();
+    final BpmnModelInstance processDefinition =
+        Bpmn.createExecutableProcess(processId)
+            .startEvent("start")
+            .inclusiveGateway("fork")
+            .defaultFlow()
+            .sequenceFlowId("flow1")
+            .conditionExpression("= contains(str,\"a\")")
+            .serviceTask("task1", b -> b.zeebeJobType("type1"))
+            .inclusiveGateway("join")
+            .endEvent("end")
+            .moveToNode("fork")
+            .sequenceFlowId("flow2")
+            .conditionExpression("= contains(str,\"b\")")
+            .connectTo("join")
+            .moveToNode("fork")
+            .sequenceFlowId("flow3")
+            .conditionExpression("= contains(str,\"c\")")
+            .serviceTask("task3", b -> b.zeebeJobType("type3"))
+            .connectTo("join")
+            .done();
```

Comment:
🔧 This test case is showing lots of promise, but it can be improved upon. Please add a service task to each flow, so each path can be controlled. Then verify that the joining inclusive gateway is activated exactly once for each scenario. You're almost there, but there are a few cases missing.

Reply:
Each flow works fine with a task . At first if there is a task between the fork and join inclusive gateway, the joining inclusive gate way is activited more then once. This test is for this scenario that i fixed last day

Reply:
Each flow works fine with a task . At first if there is a task between the fork and join inclusive gateway, the joining inclusive gate way is activited more then once. This test is for this scenario.